File: leetcode/0447-number-of-boomerangs/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int numberOfBoomerangs(int[][] points) {
        int total = 0;

        for (int[] p0 : points) {
            Map<Integer, Integer> distCount = new HashMap<>();
            for (int[] p1 : points) {
                int dx = p1[0] - p0[0];
                int dy = p1[1] - p0[1];
                int d = dx * dx + dy * dy;
                distCount.merge(d, 1, Integer::sum);
            }

            for (int count : distCount.values()) {
                total += count * (count - 1);
            }
        }

        return total;
    }
}

File: leetcode/0447-number-of-boomerangs/solution.py
from typing import List


class Solution:
    def numberOfBoomerangs(self, points: List[List[int]]) -> int:
        total = 0

        for x0, y0 in points:
            dist_count = {}
            for x1, y1 in points:
                d = (x1 - x0) ** 2 + (y1 - y0) ** 2
                dist_count[d] = dist_count.get(d, 0) + 1

            for count in dist_count.values():
                total += count * (count - 1)

        return total
